### PR TITLE
samples: mesh: enable SMP service for DFU [v1.14.0 milestone]

### PR DESCRIPTION
**Pull request by vikrant8052**
This commit enable Bluetooth SMP Service which will help
us to do Device Firmware Upgrade over thr air. By default
it is not enable.

Reference: $zephyr/samples/subsys/mgmt/mcumgr/smp_svr

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

### CODE REVIEW

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/prj.conf
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/prj.conf]: It seems like that is the opposite of what setting this to `n` accomplishes. What is the goal here?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/common.h
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/common.h]: This should be an application level Kconfig option, no? And in fact, why not just key it off of whether or not MCUMGR is enabled?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/prj.conf
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/prj.conf]: Why is this being set to `n` if an MCUboot compatible binary is desired? That will have the opposite effect.

In general I think if you want to get this in for v1.14 your best hope is to split this all out into a separate config fragment that can get pulled in with CONF_FILE. See https://docs.zephyrproject.org/latest/application/index.html#application-configuration for details.

Even that seems like it might be a bit of a stretch, I am not sure what the maintainers of this app think.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: This file should added to the build using `zephyr_sources_ifdef` using an appropriate Kconfig option for if the smp_svr backend was enabled, rather than unconditionally building it and having it sometimes do nothing.

**Review by mbolivar — CHANGES_REQUESTED** [review]

**Comment by vikrant8051** [comment]
@mbolivar 
Thanks for initiating the review process. 

As of now, implementing things as per your suggestions is beyond my scope.

So please bare me !!

I don't wanna make this App by default compatible for #mcuboot 
User should simply build & run it without prior flashing of bootloader.

**Comment by vikrant8051** [comment]
I've tried to improve the things as per my understanding. Please consider those changes.

But I am not able to compile the code if
1) I remove '#" in front of **list(APPEND CONF_FILE prj_smp_svr.conf)** in CMakeLists.txt
2) add/cooment-out  #define ENABLE_SMP_SVR in common.h

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt]: This looks like a debug line which is not needed.

My point is that the user can use `-DOVERLAY_CONFIG=prj_smp_svr.conf` at cmake time to mix in a fragment which will make the application work under MCUboot and bring the OTA update code into the build.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: I think this should probably be done using `SYS_INIT()` at application priority in src/smp_svr.c instead. That way it is done if and only if necessary and the main file doesn't need the ifdef.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: Is this needed? I don't know how mcumgr works in detail but I would imagine that there might be away to repeatedly queue work to do at timer expiration from the other file as well. Or does the system work queue not have enough stack and so you're putting this on the main thread or something?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: There's no need for this AFAICT. The file is only compiled if it is needed, i.e. if `CONFIG_BOOTLOADER_MCUBOOT=y`.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: Just looking at this command, it doesn't seem like it can't be done on the system work queue. If this is just updating some counters, it shouldn't take much time or consume a lot of stack.

**Review by mbolivar — CHANGES_REQUESTED** [review]

**Comment by mbolivar** [comment]
> @mbolivar
> Thanks for initiating the review process.

No problem.

This is not my area in Zephyr but I've tried to leave you a few more hints.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt
Comment (nvlsianpu) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt]: why not depends on CONFIG_MCUMGR istead

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: I am too unaware about background implementation of mcumgr & have simply followed the coding style mentioned in samples/subsys/mgmt/mcumgr/smp_svr.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: It is as per samples/subsys/mgmt/mcumgr/smp_svr.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt]: done !!

**Comment by vikrant8051** [comment]
if we include smp_svr.c in CMakeLists.txt
 `zephyr_sources_ifdef(CONFIG_MCUMGR src/smp_svr.c) `
like this way then I can't build the app.

**Comment by vikrant8051** [comment]
That is why content under smp_svr.c is under 

#if defined(CONFIG_MCUMGR)

#endif

**Comment by vikrant8051** [comment]
@nvlsianpu @mbolivar 
With current changes only i can build the App successfully.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: I've replaced while (1) { ... k_sleep(1000); ...} with 
`k_timer_start(&smp_svr_timer, K_MSEC(1000), 0);`

And even after that things are working as expected.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: https://github.com/zephyrproject-rtos/zephyr/pull/13135#discussion_r255399490

**Comment by vikrant8051** [comment]
@jhedberg would you like to add something from your side as correction ?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: This is not how zephyr manages whether sources get compiled. This must be done in the build system by using zephyr_sources_ifdef(), zephyr_sources_if_kconfig(), or similar.

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: Yes, I tried this. But as I already mentioned in last some comments that after it we can't build the app. Some include files are not get properly linked & this is cause behind it.

Comment (mbolivar) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: Then you need to figure out why 😉

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/smp_svr.c]: Finally, I have figured out !!!!

**Comment by vikrant8051** [comment]
@jhedberg @carlescufi @mbolivar 
App building issue has resolved. Now it is time to check the implementation. Kindly add your valuable feedback on it.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt
Comment (mbolivar) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/CMakeLists.txt]: Please use `target_sources_ifdef(CONFIG_MCUMGR app PRIVATE src/smp_svr.c)` instead.

**Review by mbolivar — DISMISSED** [review]
Structurally I just see one more issue with this PR. After that is resolved I think it is up to the application maintainer.

**Comment by vikrant8051** [comment]
@mbolivar Done !! And thanks for more appropriate suggestion.

**Comment by mbolivar** [comment]
@jhedberg @nvlsianpu I see no obvious issues with this pull request, so I've dismissed -1 and will leave it up to you from here.

**Comment by nvlsianpu** [comment]
I see no obstacles for this PR merge.
@jhedberg - this is mesh example, so pleas look at it.

**Comment by jhedberg** [comment]
> @jhedberg - this is mesh example, so pleas look at it.

@nvlsianpu haven't forgotten about this, but since it's not a bug fix I think it may need to wait until 1.14 is released.

**Comment by nvlsianpu** [comment]
^^ yes.

**Review by jhedberg — APPROVED** [review]

**Comment by zakariae87** [comment]
This sample doesnt build corrcectly under zephyr 2.1.99

**Comment by jhedberg** [comment]
> This sample (/zephyrproject/zephyr/samples/boards/nrf/mesh/onoff_level_lighting_vnd_app) doesnt build corrcectly under zephyr 2.1.99

@zakariae87 I just tried building it with latest master and had no problems (for board nrf52840_pca10056)

**Pull request merged**
